include/solvers/mpm_base.tcc

```diff
@@ -1537,4 +1531,48 @@ void mpm::MPMBase<Tdim>::initialise_linear_solver(
             std::shared_ptr<mpm::SolverBase<Eigen::SparseMatrix<double>>>>(
             dof, lin_solver));
   }
+}
+
+//! Initialise nonlocal mesh
+template <unsigned Tdim>
+void mpm::MPMBase<Tdim>::initialise_nonlocal_mesh(const Json& mesh_props) {
+  //! Shape function name
+  const auto cell_type = mesh_props["cell_type"].template get<std::string>();
+  try {
+    if (cell_type.back() == 'B') {
```

Comment:
Not sure checking the last letter of `"cell_type"` is very robust or intuitive. Was there a specific reasons you didn't just check the whole string?

Reply:
Good question! Here are my initial thoughts:
There is no specific reason, actually, other than it will be a bit messy otherwise. For the current case, we need to check 2D and 3D elements, i.e. "ED2Q4P2B" and "ED3H8P2B". Think about if one wants to expand the BSpline code to order 3, then there will be other types for different dimensions. How about if there are other nonlocal shape functions, e.g. LME, CPDI?

If each of the shape function types needs to have 2 at least, we will have quite a massive if statement.

Since the functions for the same shape func type are generally the same. I am trying to make it a convention of the last letters of nonlocal shape func, e.g.:
- "B" for BSpline 
- "L" for Local Maximum Entropy
We only then need to distinguish the treatment to initialize each of these different categories.

At least that's neater in my opinion. Any other thoughts on this will be greatly appreciated.

Reply:
I follow. Using 
* "B" for Bspline
* "L" for LME

seems reasonable since, as you say, there will be the only 2 catagories to distinguish.